tests/src/test_datafile.cpp

```diff
@@ -0,0 +1,270 @@
+/* test_datafile.cpp
+Copyright (c) 2022 by quyykk
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "es-test.hpp"
+
+// Include only the tested class's header.
+#include "../../source/DataFile.h"
+
+// Include a helper functions.
+#include "datanode-factory.h"
+#include "output-capture.hpp"
+#include "../../source/text/Format.h"
+
+// ... and any system includes needed for the test file.
+#include <algorithm>
+#include <iterator>
+#include <sstream>
+#include <string>
+#include <vector>
+
+namespace { // test namespace
+// #region mock data
+
+// Insert file-local data here, e.g. classes, structs, or fixtures that will be useful
+// to help test this class/method.
+const std::string missingQuoteWarning = "Warning: Closing quotation mark is missing:";
+const std::string mixedCommentWarning = "Warning: Mixed whitespace usage for comment at line";
+const std::string mixedNodeWarning = "Warning: Mixed whitespace usage at line";
+
+std::vector<std::string> Split(const std::string &str)
+{
+	auto result = Format::Split(str, "\n");
+
+	// Strip any empty lines from the output.
+	result.erase(std::remove_if(result.begin(), result.end(), [](const std::string &val)
+		{
+			return val.empty();
+		}), result.end());
+
+	return result;
+}
+
+// #endregion mock data
+
+
+
+// #region unit tests
+SCENARIO( "Creating a DataFile", "[DataFile]") {
+	GIVEN( "A default constructed DataFile" ) {
+		const DataFile root;
+		THEN( "it has the correct default properties" ) {
+			CHECK( root.begin() == root.end() );
+		}
+	}
+	GIVEN( "A DataFile created with a stream" ) {
+		std::istringstream stream(R"(
+node1
+	something
+node2 hi
+	something else
+)");
+		const DataFile root(stream);
+
+		THEN( "it has the correct properties" ) {
+			REQUIRE( std::distance(root.begin(), root.end()) == 2 );
+			CHECK( root.begin()->Token(0) == "node1" );
+			CHECK( std::next(root.begin())->Token(0) == "node2" );
+		}
```

Comment:
A granular approach here would be something like "iterating visits each node that has no indentation prefix"

Even in this one-shot approach, you're missing tests for what happens to nodes that do have an indentation prefix (beyond the implicit test that they are not directly iterated), and also a test when the first character is `#`.

Reply:
Done, but your next review is probably that those tests aren't granular enough. TBH I think it's fine for this case.